feat: `utilityGetUtilityContext` oracle

benesjan opened the pull request:
Introduces a new `utilityGetUtilityContext` oracle that returns all the info needed for `UtilityContext` construction in a single call. This reduces the number of `AztecNode` calls from 4 to 1. This should significantly speed up utility function execution when there is a high latency between PXE and the node.

## Bug fix in this PR
Before, in PXE, there was a serious bug in the UtilityContext construction because it was constructed based on values obtained from aztec node. This is just wrong because the simulation should be run against the PXE's view of the network state and not against the node's view. I think we didn't stumble upon this being an issue because we perform a sync right when the simulation is started. But I think it would be inevitable to happen that a node would sync a block from p2p in between the initial PXE sync and construction of the utility context and at that point we would enter an undefined behavior and a world of pain of debugging.

## Followup work
As Nico mentioned in review round 1 seems to make sense to nuke the following oracles:
- `utilityGetChainId`
- `utilityGetVersion`
- `utilityGetTimestamp`

and possibly also 
- `utilityGetBlockNumber`

as all these values are contained inside the response from the `utilityGetUtilityContext` oracle. That would also allow me to nuke these handlers from `PXEOracleInterface` which would make me very happy because the naming there was terrible (`PXEOracleInterface::getBlockHeader` returned block header of the latest block synchronized by `PXE` while `getTimestamp` and `getBlockNumber` returned values from aztec node.

benesjan commented:
* **#16450** <a href="https://app.graphite.dev/github/pr/AztecProtocol/aztec-packages/16450?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#16179** <a href="https://app.graphite.dev/github/pr/AztecProtocol/aztec-packages/16179?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈 <a href="https://app.graphite.dev/github/pr/AztecProtocol/aztec-packages/16179?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#16132** <a href="https://app.graphite.dev/github/pr/AztecProtocol/aztec-packages/16132?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#16162** <a href="https://app.graphite.dev/github/pr/AztecProtocol/aztec-packages/16162?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* `next`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->

benesjan commented on yarn-project/pxe/src/contract_function_simulator/pxe_oracle_interface.ts:
```diff
@@ -256,16 +257,26 @@ export class PXEOracleInterface implements ExecutionDataProvider {
-  public async getChainId(): Promise<number> {
-    return await this.aztecNode.getChainId();
+  public getChainId(): Promise<number> {
+    return this.aztecNode.getChainId();
```
Pulled the Jan here 😜

It bother me to have there the unnecessary async and await.

benesjan commented on yarn-project/pxe/src/contract_function_simulator/pxe_oracle_interface.ts:
```diff
@@ -256,16 +257,26 @@ export class PXEOracleInterface implements ExecutionDataProvider {
    */
-  public async getVersion(): Promise<number> {
-    return await this.aztecNode.getVersion();
+  public getVersion(): Promise<number> {
```
Pulled the Jan here 😜

It bother me to have there the unnecessary async and await.

Thunkar commented on yarn-project/stdlib/package.json:
```diff
@@ -33,6 +33,7 @@
     "./tx": "./dest/tx/index.js",
     "./fees": "./dest/fees/index.js",
     "./note": "./dest/note/index.js",
+    "./oracle": "./dest/oracle/index.js",
```
It bothers me slightly that both this and the node get "contaminated". Can we come up with a name for the `UtilityContextWithoutContractAddress` concept before it leaves the node (it's essentially a bunch of data related to the tip of the current pending chain) and assemble the context on PXE's side?

benesjan replied:
It also came to my mind and I was trying to figure out a name for it but did not come up with anything better. Think it's better to leave with the current name as it's descriptive and I think just introducing some fancy name for this one thing wouldn't be helpful as it's not used for anything else other than constructing a `UtilityContext`.

Agree that the name is ugly though 🤣

benesjan replied:
Good hunch Grego. Nico made me realize that I was genuinely just stupid here as I can simply obtain the header and construct the context based on that - no new node endpoints necessary. Reworked the whole PR and now there are also no ugly exports from `stdlib`.

nventuro commented on yarn-project/aztec-node/src/aztec-node/server.ts:
```diff
@@ -570,6 +571,24 @@ export class AztecNodeService implements AztecNode, AztecNodeAdmin, Traceable {
     return Promise.resolve(this.l1ChainId);
   }
 
+  public getUtilityContextWithoutContractAddress(): Promise<UtilityContextWithoutContractAddress> {
+    return Promise.all([this.getBlockHeader('latest'), this.getVersion(), this.getChainId()]).then(
+      ([blockHeader, version, chainId]) => {
+        if (!blockHeader) {
+          throw new Error(
+            'Aztec Node failed to obtain the latest block header when fetching the utility context without contract address.',
+          );
+        }
+        return UtilityContextWithoutContractAddress.from({
+          blockNumber: blockHeader.globalVariables.blockNumber,
+          timestamp: blockHeader.globalVariables.timestamp,
+          version,
+          chainId,
+        });
+      },
+    );
```
Unsure about this. Merging these in a single call is _fine_ I guess, but I'd definitely not pollute the node with the term 'utility context'.

Alternatively we can just do
```ts
const [ blockNumber, timestamp, version, chaind ] = await Promise.all([node.getBlockNumber(), node.getTimestamp(), node.getVersion(), node.getChainId()];
```

and get an equivalent result.

benesjan replied:
You made me realize that I was genuinely just stupid here as I can simply obtain the header and construct the context based on that - no new node endpoints necessary. Reworked the whole PR and now there are also no ugly exports from `stdlib`. Thanks

nventuro commented on noir-projects/aztec-nr/aztec/src/context/utility_context.nr:
```diff
@@ -14,31 +12,35 @@ pub struct UtilityContext {
-        let version = get_version();
-        let chain_id = get_chain_id();
-        Self { block_number, timestamp, contract_address, version, chain_id }
+        get_utility_context()
```
Why not remove the other oracles?

nventuro replied:
I mean that'd be a bigger change as it'd force us to pass the `UtilityContext` around possibly. But I'm not sure that's a bad thing.

benesjan replied:
Agree that it makes sense. Will do that in a followup PR.

nventuro commented on noir-projects/aztec-nr/aztec/src/context/utility_context.nr:
```diff
@@ -1,9 +1,7 @@
+use crate::oracle::{execution::get_utility_context, storage::storage_read};
+use protocol_types::{address::AztecAddress, traits::Packable};
 
+// If you'll modify this struct don't forget to update utility_context.ts as well.
```
This is covered by `TestEnvironment` tests regardless, where we check that utility contexts get created with the correct values. Double-check that to be sure though.

nventuro commented:
Did not approve as I'm not sure about the node change and it seems @Thunkar isn't either, and that change then resulted in a bunch of other changes (like the exports etc.).

benesjan commented on noir-projects/aztec-nr/aztec/src/messages/discovery/mod.nr:
```diff
@@ -61,10 +61,13 @@ pub struct NoteHashAndNullifier {
 
-/// Performs the message discovery process, in which private are downloaded and inspected to find new private notes,
-/// partial notes and events, etc., and pending partial notes are processed to search for their completion logs.
+/// Performs the message discovery process, in which private logs are downloaded and inspected to find new private
+/// notes, partial notes and events, etc., and pending partial notes are processed to search for their completion logs.
 /// This is the mechanism via which a contract updates its knowledge of its private state.
 ///
+/// Note that the state is synchronized up to the latest block synchronized by PXE. That should be close to the chain
+/// tip as block synchronization is performed before contract function simulation is done.
```
Did the change here as on the first line was a typo (missing "logs" word) + added the note as it's a valuable info to have here.

benesjan commented on noir-projects/aztec-nr/aztec/src/macros/functions/utils.nr:
```diff
@@ -288,10 +288,9 @@ pub(crate) comptime fn transform_utility(f: FunctionDefinition) {
         quote { let mut context = dep::aztec::context::utility_context::UtilityContext::new(); };
-    let module_has_storage = module_has_storage(f.module());
 
     // Initialize Storage if module has storage
-    let storage_init = if module_has_storage {
+    let storage_init = if module_has_storage(f.module()) {
```
Couldn't help not doing this unrelated change when I was investigating here if message discovery call is being run against the same block header from which the utility context was created.

Thunkar reviewed (approved):
Nice!

benesjan committed:
feat: `utilityGetUtilityContext` oracle

Introduces a new `utilityGetUtilityContext` oracle that returns all the info needed for `UtilityContext` construction in a single call. This reduces the number of `AztecNode` calls from 5 to 1. This should significantly speed up utility function execution when there is a high latency between PXE and the node.

## Bug fix
Before this PR there was a serious bug in the `UtilityContext` construction because it was constructed based on values obtained from aztec node. This is just wrong because the simulation should be run against the `PXE`'s view of the network state and not against the node's view. I think we didn't stumble upon this being an issue because we perform a sync right when the simulation is started. But I think it would be inevitable to happen that a node would sync a block from p2p in between the initial `PXE` sync and construction of the utility context and at that point we would enter an undefined behavior and a world of pain of debugging.

## Followup work
As Nico mentioned in review round 1 seems to make sense to nuke the following oracles:
- `utilityGetChainId`
- `utilityGetVersion`
- `utilityGetTimestamp`

and possibly also
- `utilityGetBlockNumber`

as all these values are contained inside the response from the `utilityGetUtilityContext` oracle. That would also allow me to nuke these handlers from `PXEOracleInterface` which would make me very happy because the naming there was terrible (`PXEOracleInterface::getBlockHeader` returned block header of the latest block synchronized by `PXE` while `getTimestamp` and `getBlockNumber` returned values from aztec node.
